optuna/_transform.py

```diff
@@ -0,0 +1,271 @@
+) -> Tuple[numpy.ndarray, List[numpy.ndarray]]:
+    assert len(search_space) > 0, "Cannot transform if no distributions are given."
+
+    n_bounds = sum(
+        len(d.choices) if isinstance(d, CategoricalDistribution) else 1
+        for d in search_space.values()
+    )
+
+    bounds = numpy.empty((n_bounds, 2), dtype=numpy.float64)
+    column_to_encoded_columns: List[numpy.ndarray] = []
+
+    bound_idx = 0
+    for distribution in search_space.values():
+        d = distribution
+        if isinstance(d, CategoricalDistribution):
+            n_choices = len(d.choices)
+            bounds[bound_idx : bound_idx + n_choices] = (0, 1)  # Broadcast across all choices.
+            column_to_encoded_columns.append(numpy.arange(bound_idx, bound_idx + n_choices))
+            bound_idx += n_choices
+        elif isinstance(
+            d,
+            (
+                UniformDistribution,
+                LogUniformDistribution,
+                DiscreteUniformDistribution,
+                IntUniformDistribution,
+                IntLogUniformDistribution,
+            ),
+        ):
+            if isinstance(d, UniformDistribution):
+                bds = (
+                    _transform_numerical_param(d.low, d, transform_log),
+                    _transform_numerical_param(d.high, d, transform_log),
+                )
+            elif isinstance(d, LogUniformDistribution):
+                bds = (
+                    _transform_numerical_param(d.low, d, transform_log),
+                    _transform_numerical_param(d.high, d, transform_log),
+                )
+            elif isinstance(d, DiscreteUniformDistribution):
+                half_step = 0.5 * d.q if transform_step else 0.0
+                bds = (
+                    _transform_numerical_param(d.low, d, transform_log) - half_step,
+                    _transform_numerical_param(d.high, d, transform_log) + half_step,
+                )
+            elif isinstance(d, IntUniformDistribution):
+                half_step = 0.5 * d.step if transform_step else 0.0
+                bds = (
+                    _transform_numerical_param(d.low, d, transform_log) - half_step,
+                    _transform_numerical_param(d.high, d, transform_log) + half_step,
+                )
+            elif isinstance(d, IntLogUniformDistribution):
+                half_step = 0.5 if transform_step else 0.0
+                bds = (
+                    _transform_numerical_param(d.low - half_step, d, transform_log),
+                    _transform_numerical_param(d.high + half_step, d, transform_log),
+                )
+            else:
+                assert False, "Should not reach. Unexpected distribution."
+
+            bounds[bound_idx] = bds
+            column_to_encoded_columns.append(numpy.atleast_1d(bound_idx))
+            bound_idx += 1
+        else:
+            assert False, "Should not reach. Unexpected distribution."
```

Comment:
How about testing to reach this `False` assertion with an unexpected distribution?

Reply:
We'll be testing the internals of something that's already internal. Seems a bit overkill to be honest.

Reply:
I'm not sure about this idea, but shouldn't we add the test to increase test coverage?

Reply:
Hm, yes but this is not just any code but code that should never be reached unless internals are broken 🤔

Reply:
Yes, you are right. This code should never be reached. I'm not intended to insist on my opinion, so let me resolve this conversation.